RecoHI/HiEvtPlaneAlgos/src/EvtPlaneProducer.cc

```diff
@@ -214,92 +365,92 @@ EvtPlaneProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
-	rp[etHFm5    ]->addParticle (tower_energyet,   s15,   s25,   tower_eta);
-
-	rp[etHF6     ]->addParticle (tower_energyet,   s16,   s26,   tower_eta);
-	rp[etHFp6    ]->addParticle (tower_energyet,   s16,   s26,   tower_eta);    
-	rp[etHFm6    ]->addParticle (tower_energyet,   s16,   s26,   tower_eta);
+	double minet = minet_;
+	double maxet = maxet_;
+	for(int i = 0; i<NumEPNames; i++) {
+	  if(minet_<0) minet = minTransverse[i];
+	  if(maxet_<0) maxet = maxTransverse[i];
+	  if(tower_energyet<minet) continue;
+	  if(tower_energyet>maxet) continue;
+	  if(EPDet[i]==HF) {
+	    double w = tower_energyet;
+	    if(loadDB_) w = tower_energyet*flat[i]->GetEtScale(vzr_sell,bin);
+	    if(EPOrder[i]==1 ) {
+	      if(MomConsWeight[i][0]=='y' && loadDB_ ) {
+		w = flat[i]->GetW(tower_energyet, vzr_sell, bin);
+	      }
+	      if(tower_eta<0 ) w=-w;
+	    }
+	    rp[i]->addParticle(w,tower_energyet,sin(EPOrder[i]*tower_phi),cos(EPOrder[i]*tower_phi),tower_eta);
+	  }
+	}
       } 
     }
+
+    //Castor part
+    
+ 
+    iEvent.getByToken(castorToken,castorCollection_);
+    
+    if(castorCollection_.isValid()){
+      for (std::vector<reco::CastorTower>::const_iterator j = castorCollection_->begin();j !=castorCollection_->end(); j++) {   
+       	tower_eta        = j->position().eta();
+	double labang = 2*TMath::ATan(TMath::Exp(-tower_eta));
+       	tower_phi        = j->position().phi();
+       	tower_energyet     = j->energy()*TMath::Sin(labang);
+	double minet = minet_;
+	double maxet = maxet_;
+       	for(int i = 0; i<NumEPNames; i++) {
+       	  if(EPDet[i]==Castor) {
+	    if(minet_<0) minet = minTransverse[i];
+	    if(maxet_<0) maxet = maxTransverse[i];
+	    if(tower_energyet<minet) continue;
+	    if(tower_energyet>maxet) continue;
+       	    double w = tower_energyet;
+       	    if(EPOrder[i]==1 ) {
+       	      if(MomConsWeight[i][0]=='y' && loadDB_ ) {
+       		w = flat[i]->GetW(tower_energyet, vzr_sell, bin);
+       	      }
+       	      if(tower_eta<0 ) w=-w;
+       	    }
+       	    rp[i]->addParticle(w,tower_energyet,sin(EPOrder[i]*tower_phi),cos(EPOrder[i]*tower_phi),tower_eta);
+       	  }
+       	}
+      } 
+    }
+
     //Tracking part
     
     double track_eta;
     double track_phi;
     double track_pt;
-
-    Handle<reco::TrackCollection> tracks;
-    iEvent.getByToken(trackCollection_, tracks);
-    
-    if(tracks.isValid()){
-      for(reco::TrackCollection::const_iterator j = tracks->begin(); j != tracks->end(); j++){
-	
-	//Find possible collections with command line: edmDumpEventContent *.root
-	
-	edm::Handle<reco::VertexCollection> vertex;
-	iEvent.getByToken(vtxCollection_, vertex);
-	
-	// find the vertex point and error
-
+   
+    iEvent.getByToken(trackToken, trackCollection_);
+    if(trackCollection_.isValid()){
+      for(reco::TrackCollection::const_iterator j = trackCollection_->begin(); j != trackCollection_->end(); j++){	
+	iEvent.getByLabel(vertexTag_, vertex_);	    
```

Comment:
- use ```iEvent.getByToken(vertexToken, vertex_);	``` instead
- move reading of the vertex_ handle outside of the loop over tracks